File: 10.regular-expression-matching.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int i = 0, j = 0;
        while(j < p.length()) {
            if (j + 1 < p.length() && p[j + 1] == '*') {
                if (isMatch(s.substr(i), p.substr(j + 2))) {
                    return true;
                }
                int k = 0;
                while (i + k < s.length() && (s[i + k] == p[j] || p[j] == '.')) {
                    if (isMatch(s.substr(i + k + 1), p.substr(j + 2))) {
                        return true;
                    }
                    k++;
                }
                return false;
            } else if (i < s.length() && (p[j] == s[i] || p[j] == '.')) {
                    i++;
                    j++;
            } else {
                return false;
            }
        }
        return i == s.length();
    }
};

File: 10.regular-expression-matching.py
class Solution:
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        i, j = 0, 0
        while i < len(p):
            if j + 1 < len(p) and p[j + 1] == '*':
                if self.isMatch(s[i:], p[j + 2:]):
                    return True
                k = 0
                while i + k < len(s) and (s[i + k] == p[j] or p[j] == '.'):
                    if self.isMatch(s[i + k + 1:], p[j + 2:]):
                        return True
                    k += 1
                return False
            elif i < len(s) and (p[j] == s[i] or p[j] == '.'):
                i += 1
                j += 1
            else:
                return False
        return i == len(s)
